cpp/ql/lib/semmle/code/cpp/ir/implementation/raw/internal/TranslatedElement.qll

```diff
@@ -255,6 +257,39 @@ private predicate usedAsCondition(Expr expr) {
   )
 }
 
+private predicate hasThrowingChild(Expr e) {
+  e = any(ThrowExpr throw).getFullyConverted()
+  or
+  exists(Expr child |
+    e = getRealParent(child) and
+    hasThrowingChild(child)
+  )
+}
+
+private predicate isInConditionalEvaluation(Expr e) {
+  exists(ConditionalExpr cond |
+    e = cond.getThen().getFullyConverted() and not cond.isTwoOperand()
+    or
+    e = cond.getElse().getFullyConverted()
+    or
+    // If one of the operands throws then the temporaries constructed in either
+    // branch will also be attached to the ternary expression. We suppress
+    // those destructor calls as well.
+    hasThrowingChild([cond.getThen(), cond.getElse()]) and
+    e = cond.getFullyConverted()
+  )
+  or
+  e = any(LogicalAndExpr lae).getRightOperand().getFullyConverted()
+  or
+  e = any(LogicalOrExpr loe).getRightOperand().getFullyConverted()
+  or
+  isInConditionalEvaluation(getRealParent(e))
+}
+
+private predicate isConditionalTemporaryDestructorCall(DestructorCall dc) {
+  isInConditionalEvaluation(dc.getQualifier().(ReuseExpr).getReusedExpr())
```

Comment:
Shouldn't this check that it's actually a temporary that is being re-used?

Reply:
Yeah, it probably should. Isn't that always the case currently, though?

Reply:
In any case, I've added https://github.com/github/codeql/pull/16125/commits/c325a79206db40714bc0aee126cf62e446b9369e to make this explicit. It's probably best to check for this explicitly in case we start generating `ReuseExpr` for expressions that aren't temporaries

Reply:
Yes, that's more by accident, because of the way re-use expressions are attached to `delete`s. I think the thing that really needs to be checked here is that it's an implicit call attached to an expression? The expression part is obvious, and the temporary part is really irrelevant.

Reply:
So https://github.com/github/codeql/commit/c325a79206db40714bc0aee126cf62e446b9369e is indeed what we want.

Reply:
Oh, right. Thanks for checking this!